src/psyclone/psyGen.py

```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
         return Literal(node.items[0], parent=parent)
 
 
+class Symbol(object):
+    '''
+    Symbol item for the Symbol Table. It contains information about: the name
+    of the symbol, its datatype, the number of dimensions and the symbol kind
+    (e.g. local, external, read_arg, write_arg, rw_arg).
+
+    :param name: Name of the symbol.
+    :type name: string
+    :param datatype: Data type of the symbol.
+    :type datatype: string
+    :param dimensions: Dimensions of the symbol (0 represents an scalar
+                       symbol).
+    :type dimensions: list of integers
+    :param kind: Information about the variable declaration attributes.
+    :type kind: string
+    '''
+    def __init__(self, name, datatype=None, dimensions=0, kind=None):
+        self._name = name
+        self._datatype = datatype
+        self._dimensions = dimensions
+        self._kind = kind
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def datatype(self):
+        return self._datatype
+
+    @property
+    def dimensions(self):
+        return self._dimensions
+
+    @property
+    def kind(self):
+        return self._kind
+
+    def __str__(self):
+        return (self.name + "<" + self.datatype + "," + str(self.dimensions) +
+                "," + self.kind + ">")
+
+
+class SymbolTable(object):
+    '''
+    Encapsulates the symbols table and provides methods to declare new symbols
+    and look up existing symbols. It is implemented as a monolithic scope
+    symbol table.
+    '''
+    def __init__(self):
+        self._symbols = {}
+
+    def declare(self, name, datatype, dimensions, kind):
+        '''
+        Declare a new symbol in the symbols table.
+
+        :param name: Name of the symbol.
+        :type name: string
+        :param datatype: Data type of the symbol.
+        :type datatype: string
+        :param dimensions: Dimensions of the symbol (0 represents an scalar
```

Comment:
So `dimensions` can be a list of int or 0? It might be a bit nicer to make sure it's always a list, albeit empty if it's a scalar.

Reply:
Now is always a list, and I renamed it to 'shape' to make it clear. Scalars are an empty list.